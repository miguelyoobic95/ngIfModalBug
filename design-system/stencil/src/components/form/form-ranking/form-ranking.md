---
name: Ranking
category: Form Components
---


```yoo-form-ranking.html

    <yoo-form-ranking></yoo-form-ranking>
```
```yoo-form-ranking.js

    var comp = document.querySelector('yoo-form-ranking');
    comp.values = [{value: 'a'}, {value: 'b'}, {value: 'c'}, {value: 'd'}];

```

## Attributes
|Name|Type|Default|Description|
|---|---|---|---|
|`values`|`IItemRanking[]`|   |Values which need to be ranked|


Specify, a predefined rank if you need to. If the ranks defined are not correct, then they will be all reset.
Specify, an order for the unranked items with the order prop of the IItemRanking interface if you need to.

```tsx
export interface IItemRanking {
    value: string;
    rank?: number;
    order?: number;
}
```

### Order prop of ItemRanking

```yoo-form-ranking-order.html

    <yoo-form-ranking></yoo-form-ranking>
```
```yoo-form-ranking-order.js

    var comp = document.querySelector('yoo-form-ranking');
    comp.values = [{value: 'a', order: 1}, {value: 'b', order: 2}, {value: 'c', order: 3}, {value: 'd', order: 4}];

```

### Pre defined ranking 

```yoo-form-ranking-ranking.html

    <yoo-form-ranking></yoo-form-ranking>
```
```yoo-form-ranking-ranking.js

    var comp = document.querySelector('yoo-form-ranking');
    comp.values = [{rank: 1, value: 'a'}, {rank: 3, value: 'b'}, {rank: 4, value: 'c'}, {rank: 2, value: 'd'}]

```

## CSS

|Type|Name|Description|
|---|---|---|
|Accent|`accent`|yoo-form-slider accent|
|Dark|`dark`|yoo-form-slider dark|
|Success|`success`|yoo-form-slider success|
|Danger|`danger`|yoo-form-slider danger|
|Info|`info`|yoo-form-slider info|
|Warning|`warning`|yoo-form-slider warning|
|Gradient Accent|`gradient-accent`|label gradient accent|
|Gradient Danger|`gradient-danger`|label gradient danger|
|Gradient Info|`gradient-info`|label gradient info|
|Gradient Warning|`gradient-warning`|label gradient warning|
|Gradient Success|`gradient-success`|label gradient success|
|Gradient Dark|`gradient-dark`|label gradient dark|


### Styles variations 

```yoo-form-ranking-style.html

    <yoo-form-ranking class="accent"></yoo-form-ranking>
    <br/>
    <br/>
    <yoo-form-ranking class="danger"></yoo-form-ranking>
    <br/>
    <br/>
    <yoo-form-ranking class="success"></yoo-form-ranking>
    <br/>
    <br/>
    <yoo-form-ranking class="warning"></yoo-form-ranking>
    <br/>
    <br/>
    <yoo-form-ranking class="info"></yoo-form-ranking>
    <br/>
    <br/>
    <yoo-form-ranking class="dark"></yoo-form-ranking>
    <br/>
    <br/>
    <yoo-form-ranking class="gradient-accent"></yoo-form-ranking>
    <br/>
    <br/>
    <yoo-form-ranking class="gradient-danger"></yoo-form-ranking>
    <br/>
    <br/>
    <yoo-form-ranking class="gradient-success"></yoo-form-ranking>
    <br/>
    <br/>
    <yoo-form-ranking class="gradient-warning"></yoo-form-ranking>
    <br/>
    <br/>
    <yoo-form-ranking class="gradient-info"></yoo-form-ranking>
    <br/>
    <br/>
    <yoo-form-ranking class="gradient-dark"></yoo-form-ranking>
    <br/>
    <br/>

```
```yoo-form-ranking-style.js

    var comps = [...document.querySelectorAll('yoo-form-ranking')];
    comps.map(comp => {comp.values = [{value: 'a'}, {value: 'b'}, {value: 'c'}, {value: 'd'}];});

```

## Events

|Attr|Type|Description|
|---|---|---|
|`changed`|`ItemRanking[]`|Fired when the rank is modified by the user. You can use the value of the event to restaure a previous state on the ranking component.|