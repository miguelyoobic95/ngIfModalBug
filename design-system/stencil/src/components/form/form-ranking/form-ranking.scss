

@mixin class-color-applicator($class,$color) {
    :host(.#{$class}) {
        .ranking-container {
            .item-container {
                .rank-indicator {
                    background: cssvar($color);
                }
            }
        }
    }
}

:host {
    .ranking-container {
        border: 1px solid cssvar(dark-40);
        border-bottom: 0px;
        .item-container {
            border-bottom: 1px solid cssvar(dark-40);
            align-items: center;
            padding: 0.3rem 0.5rem;
            color: cssvar(dark);
            .rank-indicator {
                margin-right: 0.5rem;
                border-radius: 50%;
                background: cssvar(success);
                color: cssvar(light);
                height: 1.3rem;
                width: 1.3rem;
                text-align: center;
                font-size: 0.9rem;
            } 
        }
    }
}   

@include class-color-applicator(accent,accent);
@include class-color-applicator(danger,danger);
@include class-color-applicator(info,info);
@include class-color-applicator(dark,dark);
@include class-color-applicator(success,success);
@include class-color-applicator(warning,warning);
@include class-color-applicator(gradient-accent,gradient-accent);
@include class-color-applicator(gradient-danger,gradient-danger);
@include class-color-applicator(gradient-info,gradient-info);
@include class-color-applicator(gradient-dark,gradient-dark);
@include class-color-applicator(gradient-warning,gradient-warning);
@include class-color-applicator(gradient-success,gradient-success);
