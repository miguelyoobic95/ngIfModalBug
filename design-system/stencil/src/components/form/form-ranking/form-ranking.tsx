import { Component, Prop, Event, EventEmitter } from '@stencil/core';
import { IItemRanking } from '@shared/interfaces';

const MAX_NUMBER = 999999;
@Component({
    tag: 'yoo-form-ranking',
    styleUrl: 'form-ranking.scss',
    scoped: true
})
export class YooFormRankingComponent {

    @Prop({mutable: true}) values: IItemRanking[] = []; // list of values with their rank

    @Event() changed: EventEmitter<IItemRanking[]>; // emit the ordered list

    constructor() {
        this.values = this.organizeItems(this.values);
    }

    organizeItems(items: IItemRanking[]) {

        function compareItemsRanking(a: IItemRanking, b: IItemRanking): number {
            return a.rank === b.rank ? 0 : a.rank < b.rank ? -1 : 1;
        }

        function compareItemsOrder(a: IItemRanking, b: IItemRanking): number {
            return a.order === b.order ? 0 : a.order < b.order ? -1 : 1;
        }

        let rankedItems = items.filter(i => i.rank);
        rankedItems = rankedItems.sort(compareItemsRanking);
        rankedItems = rankedItems.map((i, k) => {
            return {
                value: i.value,
                rank: k + 1,
                ...i.order && {order: i.order}
            };
        });
        let nonRankedItems = items.filter(i => !i.rank).sort(compareItemsOrder);
        return [...rankedItems, ...nonRankedItems];
    }

    onItemClick(index: number) {
        let items = [...this.values];
        items[index] = { // to avoid changing the original value
            value: items[index].value,
            ...items[index].rank ? {} :  {rank: MAX_NUMBER},
            ...items[index].order && {order: items[index].order}
        };

        this.values = this.organizeItems(items);
        this.changed.emit(this.values);
    }

    renderItem(item: IItemRanking, index: number): JSX.Element {
        return (
            <div class="item-container" attr-layout="row" onClick={() => this.onItemClick(index)}>
                {item.rank ?
                    <div class="rank-indicator">
                        <span>{item.rank}</span>
                    </div>
                    : null}
                <span>{item.value}</span>
            </div>
        );
    }

    render(): JSX.Element {
        this.values = this.organizeItems(this.values);
        return (
            <div class="ranking-container" attr-layout="column">
                {this.values.map((i, k) => {
                    return this.renderItem(i, k);
                })}
            </div>
        );
    }
}