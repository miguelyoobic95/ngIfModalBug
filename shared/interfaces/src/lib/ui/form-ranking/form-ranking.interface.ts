
export interface IItemRanking {
    value: string;
    rank?: number;
    order?: number;
}
