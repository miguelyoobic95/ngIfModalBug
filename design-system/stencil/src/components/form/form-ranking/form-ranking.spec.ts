import { TestWindow } from '@stencil/core/testing';
import { YooFormRankingComponent } from './form-ranking';

describe('Form ranking Component', () => {

    it('Should build', async () => {
        let element = new YooFormRankingComponent();
        expect(element).toBeTruthy();
    });

    describe('ranking Items', ( () => {
        let element = new YooFormRankingComponent();

        it('should return all items as non ranked', () => {
            let items = [{value: 'a'}, {value: 'b'}, {value: 'c'}];
            let result = items;
            expect(element.organizeItems(items)).toEqual(result);
        });

        it('should return ranked items', () => {
            let items = [{value: 'a'}, {value: 'b', rank: 5}, {value: 'c', rank: 3}];
            let result = [{value: 'c', rank: 1}, {value: 'b', rank: 2}, {value: 'a'}];
            expect(element.organizeItems(items)).toEqual(result);
        });

        it('should return ranked items with zero rank', () => {
            let items = [{value: 'a'}, {value: 'b', rank: -1}, {value: 'c', rank: 3}];
            let result = [{value: 'b', rank: 1}, {value: 'c', rank: 2}, {value: 'a'}];
            expect(element.organizeItems(items)).toEqual(result);
        });

        it('should return ranked items and ordered non ranked items', () => {
            let items = [{ value: 'a', order: 2 }, { value: 'b', rank: 1, order: 3 }, { value: 'c', order: 1 }];
            let result = [{ value: 'b', rank: 1, order: 3 }, { value: 'c', order: 1 }, { value: 'a', order: 2 }];
            expect(element.organizeItems(items)).toEqual(result);
        });
    }));

    describe('Should change items ranking', () => {
        let element = new YooFormRankingComponent();
        element.changed = {
            emit: () => {}
        };
         it('should remove ranking from an item', () => {
            const changedSpy = jest.spyOn(element.changed, 'emit');
            element.values = [{value: 'b', rank: 1}, {value: 'c', rank: 2}, {value: 'a'}];
            let result = [{value: 'b', rank: 1}, {value: 'c'}, {value: 'a'}];
            element.onItemClick(1);
            expect(element.values).toEqual(result);
            expect(changedSpy).toHaveBeenCalledWith(result);
         });

         it('should add ranking to an item', () => {
            const changedSpy = jest.spyOn(element.changed, 'emit');
            element.values = [{value: 'b', rank: 1}, {value: 'c'}, {value: 'a'}];
           let result = [{value: 'b', rank: 1}, {value: 'c', rank: 2}, {value: 'a'}];
            element.onItemClick(1);
            expect(element.values).toEqual(result);
            expect(changedSpy).toHaveBeenCalledWith(result);
         });

    });

    describe('Rendering', () => {
        it('Should render with values without ranking', async () => {
            const window = new TestWindow();
            let element = await window.load({
                components: [YooFormRankingComponent],
                html: '<yoo-form-ranking></yoo-form-ranking>'
              });
            element.values = [{value: 'a'}, {value: 'b'}, {value: 'c'}, {value: 'd'}];
            await window.flush();
            expect(element).toMatchSnapshot();
        });

        it('Should render with values with ranking', async () => {
            const window = new TestWindow();
            let element = await window.load({
                components: [YooFormRankingComponent],
                html: '<yoo-form-ranking></yoo-form-ranking>'
              });
            element.values = [{value: 'a'}, {value: 'b', rank: 1}, {value: 'c', rank: 2}, {value: 'd'}];
            await window.flush();
            expect(element).toMatchSnapshot();
        });

        it('Should render with values with ranking and order', async () => {
            const window = new TestWindow();
            let element = await window.load({
                components: [YooFormRankingComponent],
                html: '<yoo-form-ranking></yoo-form-ranking>'
              });
            element.values = [{value: 'a', order: 1}, {value: 'b', rank: 1, order: 2}, {value: 'c', rank: 2, order: 3}, {value: 'd', order: 4}];
            await window.flush();
            expect(element).toMatchSnapshot();
        });

    });

    describe('Ranking elements ', () => {
        const VALUES = [{value: 'a'}, {value: 'b', rank: 2}, {value: 'c', rank: 1}, {value: 'd'}];
        let element, window;

        function findItem (el: HTMLElement, value: string): HTMLElement {
            let item = Array.from(el.querySelectorAll('.item-container')).find((i: HTMLElement) => {
                return i.querySelector('.item-container > span').textContent === value;
            });
            return <HTMLElement>item;
        }

        beforeEach(async () => {
            window = new TestWindow();
            element = await window.load({
                components: [YooFormRankingComponent],
                html: '<yoo-form-ranking></yoo-form-ranking>'
            });
            element.values = VALUES;
            window.flush();
        });

        it('Should select and unselect an element', async () => {
            let item = findItem(element, 'a');
            item.click();
            expect(element).toMatchSnapshot();
            item = findItem(element, 'a');
            item.click();
            expect(element).toMatchSnapshot();
        });

        it('Should unselect and select an element', async () => {
            let item = findItem(element, 'c');
            item.click();
            expect(element).toMatchSnapshot();
            item = findItem(element, 'c');
            item.click();
            expect(element).toMatchSnapshot();
        });
    });
});