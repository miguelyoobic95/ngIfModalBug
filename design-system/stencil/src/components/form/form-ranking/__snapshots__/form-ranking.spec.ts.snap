// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form ranking Component Ranking elements  Should select and unselect an element 1`] = `
<yoo-form-ranking
  class="hydrated"
  data-yoo-form-ranking-host=""
>
  
  <div
    attr-layout="column"
    class="ranking-container"
    data-yoo-form-ranking=""
  >
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      <div
        class="rank-indicator"
        data-yoo-form-ranking=""
      >
        <span
          data-yoo-form-ranking=""
        >
          1
        </span>
      </div>
      <span
        data-yoo-form-ranking=""
      >
        c
      </span>
    </div>
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      <div
        class="rank-indicator"
        data-yoo-form-ranking=""
      >
        <span
          data-yoo-form-ranking=""
        >
          2
        </span>
      </div>
      <span
        data-yoo-form-ranking=""
      >
        b
      </span>
    </div>
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      
      <span
        data-yoo-form-ranking=""
      >
        a
      </span>
    </div>
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      
      <span
        data-yoo-form-ranking=""
      >
        d
      </span>
    </div>
  </div>
</yoo-form-ranking>
`;

exports[`Form ranking Component Ranking elements  Should select and unselect an element 2`] = `
<yoo-form-ranking
  class="hydrated"
  data-yoo-form-ranking-host=""
>
  
  <div
    attr-layout="column"
    class="ranking-container"
    data-yoo-form-ranking=""
  >
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      <div
        class="rank-indicator"
        data-yoo-form-ranking=""
      >
        <span
          data-yoo-form-ranking=""
        >
          1
        </span>
      </div>
      <span
        data-yoo-form-ranking=""
      >
        c
      </span>
    </div>
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      <div
        class="rank-indicator"
        data-yoo-form-ranking=""
      >
        <span
          data-yoo-form-ranking=""
        >
          2
        </span>
      </div>
      <span
        data-yoo-form-ranking=""
      >
        b
      </span>
    </div>
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      
      <span
        data-yoo-form-ranking=""
      >
        a
      </span>
    </div>
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      
      <span
        data-yoo-form-ranking=""
      >
        d
      </span>
    </div>
  </div>
</yoo-form-ranking>
`;

exports[`Form ranking Component Ranking elements  Should unselect and select an element 1`] = `
<yoo-form-ranking
  class="hydrated"
  data-yoo-form-ranking-host=""
>
  
  <div
    attr-layout="column"
    class="ranking-container"
    data-yoo-form-ranking=""
  >
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      <div
        class="rank-indicator"
        data-yoo-form-ranking=""
      >
        <span
          data-yoo-form-ranking=""
        >
          1
        </span>
      </div>
      <span
        data-yoo-form-ranking=""
      >
        c
      </span>
    </div>
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      <div
        class="rank-indicator"
        data-yoo-form-ranking=""
      >
        <span
          data-yoo-form-ranking=""
        >
          2
        </span>
      </div>
      <span
        data-yoo-form-ranking=""
      >
        b
      </span>
    </div>
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      
      <span
        data-yoo-form-ranking=""
      >
        a
      </span>
    </div>
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      
      <span
        data-yoo-form-ranking=""
      >
        d
      </span>
    </div>
  </div>
</yoo-form-ranking>
`;

exports[`Form ranking Component Ranking elements  Should unselect and select an element 2`] = `
<yoo-form-ranking
  class="hydrated"
  data-yoo-form-ranking-host=""
>
  
  <div
    attr-layout="column"
    class="ranking-container"
    data-yoo-form-ranking=""
  >
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      <div
        class="rank-indicator"
        data-yoo-form-ranking=""
      >
        <span
          data-yoo-form-ranking=""
        >
          1
        </span>
      </div>
      <span
        data-yoo-form-ranking=""
      >
        c
      </span>
    </div>
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      <div
        class="rank-indicator"
        data-yoo-form-ranking=""
      >
        <span
          data-yoo-form-ranking=""
        >
          2
        </span>
      </div>
      <span
        data-yoo-form-ranking=""
      >
        b
      </span>
    </div>
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      
      <span
        data-yoo-form-ranking=""
      >
        a
      </span>
    </div>
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      
      <span
        data-yoo-form-ranking=""
      >
        d
      </span>
    </div>
  </div>
</yoo-form-ranking>
`;

exports[`Form ranking Component Rendering Should render with values with ranking 1`] = `
<yoo-form-ranking
  class="hydrated"
  data-yoo-form-ranking-host=""
>
  
  <div
    attr-layout="column"
    class="ranking-container"
    data-yoo-form-ranking=""
  >
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      <div
        class="rank-indicator"
        data-yoo-form-ranking=""
      >
        <span
          data-yoo-form-ranking=""
        >
          1
        </span>
      </div>
      <span
        data-yoo-form-ranking=""
      >
        b
      </span>
    </div>
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      <div
        class="rank-indicator"
        data-yoo-form-ranking=""
      >
        <span
          data-yoo-form-ranking=""
        >
          2
        </span>
      </div>
      <span
        data-yoo-form-ranking=""
      >
        c
      </span>
    </div>
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      
      <span
        data-yoo-form-ranking=""
      >
        a
      </span>
    </div>
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      
      <span
        data-yoo-form-ranking=""
      >
        d
      </span>
    </div>
  </div>
</yoo-form-ranking>
`;

exports[`Form ranking Component Rendering Should render with values with ranking and order 1`] = `
<yoo-form-ranking
  class="hydrated"
  data-yoo-form-ranking-host=""
>
  
  <div
    attr-layout="column"
    class="ranking-container"
    data-yoo-form-ranking=""
  >
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      <div
        class="rank-indicator"
        data-yoo-form-ranking=""
      >
        <span
          data-yoo-form-ranking=""
        >
          1
        </span>
      </div>
      <span
        data-yoo-form-ranking=""
      >
        b
      </span>
    </div>
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      <div
        class="rank-indicator"
        data-yoo-form-ranking=""
      >
        <span
          data-yoo-form-ranking=""
        >
          2
        </span>
      </div>
      <span
        data-yoo-form-ranking=""
      >
        c
      </span>
    </div>
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      
      <span
        data-yoo-form-ranking=""
      >
        a
      </span>
    </div>
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      
      <span
        data-yoo-form-ranking=""
      >
        d
      </span>
    </div>
  </div>
</yoo-form-ranking>
`;

exports[`Form ranking Component Rendering Should render with values without ranking 1`] = `
<yoo-form-ranking
  class="hydrated"
  data-yoo-form-ranking-host=""
>
  
  <div
    attr-layout="column"
    class="ranking-container"
    data-yoo-form-ranking=""
  >
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      
      <span
        data-yoo-form-ranking=""
      >
        a
      </span>
    </div>
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      
      <span
        data-yoo-form-ranking=""
      >
        b
      </span>
    </div>
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      
      <span
        data-yoo-form-ranking=""
      >
        c
      </span>
    </div>
    <div
      attr-layout="row"
      class="item-container"
      data-yoo-form-ranking=""
    >
      
      <span
        data-yoo-form-ranking=""
      >
        d
      </span>
    </div>
  </div>
</yoo-form-ranking>
`;
